### ISSUE TITLE
There is no way to serve source map files separately

### ISSUE THREAD

**Issue by dimsmol** [webpack/webpack#783]
On production source maps must be protected from unauthorised users. The simplest way to do it is to serve them separately and use any kind of authentication, restriction by IP or something of a kind.

But webpack always places source map files within output dir and uses relative paths as urls for them.

There is no way to generate source maps beyond of output dir, the only option is to specify something like `sourceMapFilename='srcmaps/[file].map'` and move `srcmaps` directory to some other place as a post-build step.

But there will be a problem with source map urls then. One can use plugin constructor arguments to override `sourceMappingURLComment`, but `[url]` will be substituted with a relative path including number of `../` which prevents possibility to prefix it with something indicating where source maps will be served from.

E.g. `SourceMapDevToolPlugin('srcmaps/[file].map', '\n//# sourceMappingURL=/secretplace/[url]')` will not work because for `a/b/c/d/e.js` sourceMappingURL will be `/secretplace/../../../../srcmaps/a/b/c/d/e.js.map` while we need `/secretplace/srcmaps/a/b/c/d/e.js.map`

Basically, we need:
- Ability to use something like `[url]` equivalent for sourceMappingURLComment, but providing a path relative to content root instead of js file location
- A way to specify output directory for source maps independent of output path for all other content (to avoid post-build moving step)
- Ideally, a little bit more convenient way to specify sourceMappingURL template than by using sourceMappingURLComment argument

P.S. I understand that source map files can be protected while still being located within common output dir, but it is dangerous, because if you forget to serve source maps located elsewhere you would just have a production server without source maps, but if you forget to protect deployed source maps you will expose sources to everyone.

**Comment by jhnns** [comment]
:+1:

**Comment by DavidSouther** [comment]
I would prefer the ability to control sourcemaps via the Source-Map header, which would also dovetail with this issue.

**Comment by nervgh** [comment]
I was going to ask same question :+1: 
Also webpack includes [footer](https://yadi.sk/i/CFMl2zhfeqZji) which I cannot turn off.

**Comment by vaidik** [comment]
+1 For this. This is actually a big issue now.

**Comment by ufukomer** [comment]
+1

**Comment by muodov** [comment]
Are there any technical difficulties related to this? Or is it a good ticket to start contributing to webpack? :)

**Comment by jacobwarduk** [comment]
Curious if there has been any work done towards this, or if there is another solution to the problem?

**Comment by ewohlken2** [comment]
No update on this? I'm having the same issue of needing an extra build step in order to put sourcemaps outside of my output directory

**Comment by vankop** [comment]
> On production source maps must be protected from unauthorised users. The simplest way to do it is to serve them separately and use any kind of authentication, restriction by IP or something of a kind.

`hidden-source-map` could help here.
Closing because of inactivity

**Comment by evilebottnawi** [comment]
@vankop I don't think it is solved...

**Comment by vankop** [comment]
@evilebottnawi regarding *serving* source map files separately - yes, it is not solved, but I don't think that there is any work from webpack core side. `hidden-source-map` is what you looking for (from webpack side)

To serve them in browser (as in common way) you need browser support to loading source maps. I don't know any browser that does it.. To serve them other way you need a tool for that (probably another idea for open source project)

**Comment by vankop** [comment]
for mapping stack trace, unfortunately, I also don't know any public instrument. I did it as a plugin for [graylog](https://www.graylog.org/) in our company, but it is not public..

**Comment by evilebottnawi** [comment]
@vankop you can setup `sentry` with any logic, for example you can store file with mapping like:
```json
{
  "app.js": "sourcemsp/app.js.map"
}

**Comment by vankop** [comment]
hm.. cool. I did something similar just automated on build.

Anyway I think it is not related to webpack core.

**Comment by evilebottnawi** [comment]
The problems is you can put source maps in other directories using source map plugin, but maybe you are right, it is out of scope webpack, anyway I want to hear other opinion here, and maybe we will repopen issue

**Comment by Rauttis** [comment]
I haven't used webpack for while so I don't know if this still works in newer versions, but I've managed to generate source maps to a different output folder by using a relative path like `../sourcemaps/[name].[chunkhash].js.map` as `sourceMapFilename`.

This of course doesn't help with the issue of serving them, since the source map url webpack adds to the bundle probably doesn't point to the place where you're actually serving them from. It would be nice if it was possible to prefix/override that generated URL somehow.

It's actually possible to load the source maps in Chromium dev tools after the fact by opening the relevant bundle in `Sources` and then right clicking the code -> `Add source map`, but this obviously isn't very optimal since it's something you have to do by hand.

**Comment by sokra** [comment]
It seems with a few small changes we can allow the `append` option to be a function (similar to other filename related option), so this could be better customized. It's running through `compilation.getPath` anyway, which supports functions.

**Comment by evilebottnawi** [comment]
Let's remove, if somebody interested - feel free to send a PR

**Comment by alexander-akait** [comment]
bump

**Comment by coder-freestyle** [comment]
Is this issue still open. Would want to try working on this. Thanks in advance!
